akka-actor/src/main/scala/akka/dispatch/BatchingExecutor.scala

```diff
@@ -89,21 +89,22 @@ private[akka] trait BatchingExecutor extends Executor {
         parentBlockContext = prevBlockContext
         try processBatch(this) catch {
           case t: Throwable ⇒
+            parentBlockContext = null
             resubmitUnbatched()
             throw t
-        } finally {
-          _tasksLocal.remove()
-          parentBlockContext = null
-        }
+        } finally _tasksLocal.remove()
+        parentBlockContext = null
       }
     }
 
     override def blockOn[T](thunk: ⇒ T)(implicit permission: CanAwait): T = {
+      // need to rescue the parent BC before re-running this Batch
+      val parent = parentBlockContext
+      parentBlockContext = null
```

Comment:
To make a clean slate: resubmitUnbatched may or may not schedule this Batch to run, so don’t retain a reference in either case.